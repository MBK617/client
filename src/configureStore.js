/**
 * @file Configure redux store
 */

import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import createReducer from './reducers';

/**
 * Configures and returns initial redux store
 *
 * @param {object} initialState The initial state of the application
 * 
 * @return {object} A redux store
 * 
 * @memberof module:Main
 *
 */
function configureStore(initialState = {}) {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    createReducer(),
    initialState,
    applyMiddleware(sagaMiddleware)
  );

  store.runSaga = sagaMiddleware.run;
  store.injectedReducers = {};
  store.injectedSagas = {};

  return store;
}

export default configureStore;