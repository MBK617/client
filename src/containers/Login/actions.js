export const logIn = (email, password) => ({ type: 'LOGIN', email, password })
export const logInSuccess = (token) => ({ type: 'LOGIN_SUCCESS', token })
export const logInFailure = (error) => ({ type: 'LOGIN_FAILURE', error })