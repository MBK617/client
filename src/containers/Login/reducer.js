import { fromJS } from 'immutable';

const initialState = fromJS({
  errorMessage: '',
  loading: false,
});

export default function counter(state = initialState, action) {
  switch (action.type) {
    case 'LOGIN':
      return state
        .set('loading', true)
        .set('errorMessage', '');
    case 'LOGIN_SUCCESS': 
      return state
        .set('loading', false);
    case 'LOGIN_FAILURE':
      return state
        .set('errorMessage', action.error)
        .set('loading', false);
    default:
      return state
  }
}
