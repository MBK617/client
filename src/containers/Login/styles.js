import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  root: {
    padding: 40,
  },
  error: {
    textAlign: 'center',
    color: '#ff0000',
    width: 250,
    marginBottom: 10,
  }
});

export default styles;