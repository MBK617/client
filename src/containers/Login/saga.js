import { logInSuccess } from "./actions";
import { put, takeLatest, all } from 'redux-saga/effects';
import { getRequestOptions, request } from "../../utils/request";

export function* makeLoginRequest(action) {
  const payload = {
    userEmail: action.email,
    userPassword: action.password
  }
  try {
    const result = yield request(`http://localhost:8080/api/public/user/login`, getRequestOptions('post', payload));
    yield put(logInSuccess(result.token));
  } catch (err) {
    //TODO: get error message and dispatch logInFailure
  }
}

export default function* rootSaga() {
  yield all([
    takeLatest('LOGIN', makeLoginRequest),
  ]);
}