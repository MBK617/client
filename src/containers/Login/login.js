import React, { useState } from 'react';
import { compose } from 'redux';
import { useDispatch, useSelector } from 'react-redux';
import { createSelector } from 'reselect';

import { Button, Text, View } from 'react-native';

import injectSaga from 'utils/redux/injectSaga';
import injectReducer from 'utils/redux/injectReducer';
import Input from 'components/Input';

import styles from './styles';
import reducer from './reducer';
import saga from './saga';
import { logIn } from './actions';

const selectLogin = createSelector(
  state => state.login,
  substate => substate.toJS()
);

const Login = () => {
  const dispatch = useDispatch();
  const { errorMessage } = useSelector(selectLogin);

  const [ email, setEmail ] = useState('');
  const [ password, setPassword ] = useState('');

  return (
    <View style={styles.root}>
      <Input value={email} onChangeText={setEmail} placeholder="Email"/>
      <Input value={password} onChangeText={setPassword} placeholder="Password"/>
      <Text style={styles.error}>{errorMessage}</Text>
      <Button title="Log In" onPress={()=>dispatch(logIn(email, password))}/>
    </View>
  );
}

const withReducer = injectReducer({ key: 'login', reducer });
const withSaga = injectSaga({ key: 'login', saga });

export default compose(withReducer, withSaga)(Login);
