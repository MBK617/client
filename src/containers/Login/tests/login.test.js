import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render, cleanup, fireEvent } from '@testing-library/react-native';
import { fromJS } from 'immutable';
import createSagaMiddleware from 'redux-saga';

import Login from '..';

const initialState = {
  login: fromJS({
    errorMessage: '',
    loading: false,
  })
};

describe('<Login />', () => {
  const sagaMiddleware = createSagaMiddleware();
  let store, rendered;

  const mockStore = (initialState) => ({ 
    ...configureStore([sagaMiddleware])(initialState), 
    injectedReducers: {}, 
    injectedSagas: {}, 
    runSaga: sagaMiddleware.run 
  });

  beforeEach( () => {
    store = mockStore(initialState);
    rendered = render(
      <Provider store={store}><Login /></Provider>
    );
  })

  afterEach(cleanup);

  it('should render correctly', () => {
    const tree = rendered.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display login form', () => {
    expect(rendered.getByPlaceholderText(/Email/i)).toBeTruthy();
    expect(rendered.getByPlaceholderText(/Password/i)).toBeTruthy();
    expect(rendered.getByText(/Log In/i)).toBeTruthy();
  });
  
  it('should dispatch login action', () => {
    const button = rendered.getByText(/Log In/i);
    fireEvent(button, 'press');
    const actions = store.getActions();
    expect(actions[0].type).toEqual('LOGIN');
  });

  it('should render error message', () => {
    store = mockStore({ login: initialState.login.set('errorMessage', 'An error occurred') });
    rendered = render(
      <Provider store={store}><Login /></Provider>
    );
    expect(rendered.getByText('An error occurred')).toBeTruthy();
  });
});