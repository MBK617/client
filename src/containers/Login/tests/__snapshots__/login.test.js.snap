// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render correctly 1`] = `
<View
  style={
    Object {
      "padding": 40,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "marginBottom": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Email"
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "marginBottom": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <Text
    style={
      Object {
        "color": "#ff0000",
        "marginBottom": 10,
        "textAlign": "center",
        "width": 250,
      }
    }
  >
    
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Log In
      </Text>
    </View>
  </View>
</View>
`;
